archunit-junit/junit5/engine/src/main/java/com/tngtech/archunit/junit/internal/filtering/AbstractTestNameFilter.java

```diff
@@ -0,0 +1,129 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.TestSource;
+import org.junit.platform.launcher.LauncherDiscoveryRequest;
+import org.junit.platform.launcher.PostDiscoveryFilter;
+
+public abstract class AbstractTestNameFilter implements TestSourceFilter {
+    private static final Collection<TestSelectorFactory> SELECTOR_FACTORIES = Arrays.asList(
+            new MethodSelectorFactory(),
+            new FieldSelectorFactory()
+    );
+    private final String discoveryFilterClassName;
+
+    public AbstractTestNameFilter(EngineDiscoveryRequest request, String discoveryFilterClassName) {
+        this.discoveryFilterClassName = discoveryFilterClassName;
+        Optional<PostDiscoveryFilter>  postDiscoveryFilter = findPostDiscoveryFilter(request);
+        Optional<PostDiscoveryFilter> replacement = postDiscoveryFilter.flatMap(this::throwingInitialize);
+        replacement.ifPresent(
+                newFilter -> postDiscoveryFilter.ifPresent(
+                        oldFilter -> replaceFilterIfNeeded(request, newFilter, oldFilter)));
+    }
+
+    public AbstractTestNameFilter(EngineDiscoveryRequest request) {
+        this(request, "");
+    }
+
+    private void replaceFilterIfNeeded(EngineDiscoveryRequest request, PostDiscoveryFilter newFilter, PostDiscoveryFilter oldFilter) {
+        if (!oldFilter.equals(newFilter)) {
+            try {
+                replaceFilter(request, oldFilter, newFilter);
+            } catch (ReflectiveOperationException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private Optional<PostDiscoveryFilter> throwingInitialize(PostDiscoveryFilter filter) {
+        try {
+            return initialize(filter);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void replaceFilter(
+            EngineDiscoveryRequest discoveryRequest,
+            PostDiscoveryFilter postDiscoveryFilter,
+            PostDiscoveryFilter replacement) throws ReflectiveOperationException {
+        LauncherDiscoveryRequest request = (LauncherDiscoveryRequest) discoveryRequest;
+        List<PostDiscoveryFilter> filters = new ArrayList<>(request.getPostDiscoveryFilters());
+        filters.remove(postDiscoveryFilter);
+        filters.add(replacement);
+        ReflectionUtils.setField(discoveryRequest, "postDiscoveryFilters", filters);
```

Comment:
This feels wrong to me :thinking: The JUnit Platform should provide extension points to filter without the need to break encapsulations of the platform in such extreme ways (even having to break the `final` modifier)

Reply:
This `replaceFilter` will only be used for Gradle. I can remove it for now. 

I can explain the need for this workaround and how what we could do to get rid of it in the Gradle PR, would that be okay with you?